swarm/src/lib.rs

```diff
@@ -1508,24 +1527,79 @@ impl<TBehaviour> SwarmBuilder<TBehaviour>
 where
     TBehaviour: NetworkBehaviour,
 {
-    /// Creates a new [`SwarmBuilder`] from the given transport, behaviour, local peer ID and
-    /// executor. The `Swarm` with its underlying `Network` is obtained via
-    /// [`SwarmBuilder::build`].
-    pub fn with_executor(
+    /// Private convenience function.
+    /// Creating a new [`SwarmBuilder`] from the given transport, behaviour and local peer ID.
+    fn new(
         transport: transport::Boxed<(PeerId, StreamMuxerBox)>,
         behaviour: TBehaviour,
         local_peer_id: PeerId,
-        executor: impl Executor + Send + 'static,
     ) -> Self {
         Self {
             local_peer_id,
             transport,
             behaviour,
-            pool_config: PoolConfig::new(Some(Box::new(executor))),
+            pool_config: PoolConfig::new(None),
             connection_limits: Default::default(),
         }
     }
 
+    /// Private convenience function.
+    /// Sets executor of this [`SwarmBuilder`] to a given value.
+    fn executor(mut self, executor: impl Executor + Send + 'static) -> Self {
+        self.pool_config = PoolConfig::new(Some(Box::new(executor)));
+        self
+    }
```

Comment:
Do we need this? There shouldn't be a need to set the executor past constructing the builder.

Reply:
Strictly speaking, we don't. This is a private convenience function; the end user will never see it. I can ofc do the construction in place, but this will lead to a lot of code duplication.

Reply:
Unless I am missing something, this field should only be set in two places:

1. A constructor that sets an executor
2. A constructor that doesn't set an executor

Shouldn't it be possible to implement everything else on top of these? Two (different) usages of the same field isn't critical code duplication to me.

Reply:
fixed